### PR TITLE
Developers and costs check

### PR DESCRIPTION
**Pull request by whiterabbit1983**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Add usage cost monthly aggregation and view in database
  - New migrations for monthly cost tracking and developer view

- Implement usage cost limit middleware for free tier users
  - Middleware blocks requests if cost limit exceeded or account inactive
  - Paid users bypass cost limit check

- Add and enhance usage cost calculation and fallback logic
  - Average cost fallback for unknown models
  - Expose usage cost query in API

- Add comprehensive tests for usage cost and middleware
  - Test cost limit, paid tag, error handling, and edge cases

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td><strong>developer_id.py</strong><dd><code>Remove developer active check from dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-6223236a2e515b6140ff389ecd0cab0eb37b1f19cafc52bce8ac7ecacebb1182">+1/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>env.py</strong><dd><code>Add free tier cost limit environment variable</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-1f8c23dc72151bd249217763369b0b8837ce35a7bd5bedd304996b3e828f0493">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>__init__.py</strong><dd><code>Export get_usage_cost in usage queries</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-28ce8d266837604fe3586e802c64d1c79336670f9549dfd3f950fa3a7214871c">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>create_usage_record.py</strong><dd><code>Add average cost fallback for unknown models</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-c3c3933826f43f77d2313c0c3ef45a041fb709da69dd9d3eff344a7960652e06">+40/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>get_user_cost.py</strong><dd><code>Add get_usage_cost query for developer monthly cost</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-ab11c72700686d36116dc1218c3c9337b560ac62d517541d9917f4c23ca967c8">+52/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>web.py</strong><dd><code>Add usage cost limit middleware for free users</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-6f6412b372aa58ba2fed61beed56f6a884c1797c8a91c0d8d9b715fa5c14fccf">+84/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>000038_usage_cost_monthly.down.sql</strong><dd><code>Add migration to drop monthly usage cost aggregates/views</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-fd909c2f2b8862804f86508d39adb4ed323b5f5cc62cfcd233f7b4ed85574dfc">+12/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>000038_usage_cost_monthly.up.sql</strong><dd><code>Add migration for monthly usage cost aggregation and view</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-e513fe50e851718d61b087df51d9af0d6e6b03a73160eda6be320b27524d1333">+34/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Formatting</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>get_developer.py</strong><dd><code>Minor SQL formatting update</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-c9425aed2b6db3867a3b82ba0534930f8fd5a758a4213253d7c63dd2948915a4">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>test_developer_queries.py</strong><dd><code>Remove dependency test for inactive developer</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-5b526b150628811f95686e39efec3e1f66a65ce2aa2ad54b4c4bc41f607323c0">+0/-39</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_middleware.py</strong><dd><code>Add comprehensive tests for usage cost middleware</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-dc36c95a1896a611bd298b4db67de03c5beb54bc6f966ca08196932e27ce24b4">+464/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_usage_cost.py</strong><dd><code>Add tests for usage cost query and aggregation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-9dc574814040bc04056041841663616c3de9894935338fdd91acb9ad51a44016">+276/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_usage_tracking.py</strong><dd><code>Update tests for fallback cost calculation logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1351/files#diff-c6a3147204b376524acc103abd218636bc42fae3be3360292ff9b1ade155e84b">+11/-3</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

----

> [!IMPORTANT]
> Adds monthly usage cost tracking, middleware for cost limit enforcement, and comprehensive tests for free-tier users.
> 
>   - **Behavior**:
>     - Adds monthly usage cost aggregation and view in the database with migrations `000038_usage_cost_monthly.up.sql` and `000038_usage_cost_monthly.down.sql`.
>     - Implements `usage_check_middleware` in `web.py` to block requests for free-tier users exceeding cost limits or inactive accounts.
>     - Paid users bypass cost checks.
>   - **Models**:
>     - Adds `get_usage_cost` query in `get_user_cost.py` to retrieve developer monthly costs.
>     - Updates `create_usage_record.py` to include average cost fallback for unknown models.
>   - **Environment**:
>     - Adds `free_tier_cost_limit` environment variable in `env.py`.
>   - **Tests**:
>     - Adds tests for usage cost and middleware in `test_usage_cost.py`, `test_middleware.py`, and `test_usage_tracking.py`.
>     - Removes inactive developer dependency test in `test_developer_queries.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for 03173f177842272ce22c1d5dad15d7567edabe79. You can [customize](https://app.ellipsis.dev/julep-ai/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/agents_api/queries/usage/get_user_cost.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from beartype import beartype
+
+from ...common.utils.db_exceptions import common_db_exceptions
+from ...metrics.counters import query_metrics
+from ..utils import pg_query, rewrap_exceptions, wrap_in_class
+
+usage_query = """
+SELECT developer_id, active, tags, monthly_cost AS cost
+FROM usage_cost_monthly
```
Comment (creatorrr) [inline agents-api/agents_api/queries/usage/get_user_cost.py:17]: This view should be named developer_cost_monthly

Comment (creatorrr) [reply agents-api/agents_api/queries/usage/get_user_cost.py:17]: Also, need to order by the monthly bucket descending

**Comment by creatorrr** [comment]
lgtm

**Comment by creatorrr** [comment]
Lgtm. Merged #1352  into this so you can test it @whiterabbit1983
